cpp/src/rolling/lead_lag_nested_detail.cuh

```diff
@@ -0,0 +1,227 @@
+ * @tparam InputType The datatype of the input column being aggregated
+ * @tparam PrecedingIterator Iterator-type that returns the preceding bounds
+ * @tparam FollowingIterator Iterator-type that returns the following bounds
+ * @param[in] input Nested-type input column for LEAD/LAG calculation
+ * @param[in] default_outputs Default values to use as outputs, if LEAD/LAG
+ *                            offset crosses column/group boundaries
+ * @param[in] preceding Iterator to retrieve preceding window bounds
+ * @param[in] following Iterator to retrieve following window bounds
+ * @param[in] offset Lead/Lag offset, indicating which row after/before
+ *                   the current row is to be returned
+ * @param[in] stream CUDA stream for device memory operations/allocations
+ * @param[in] mr device_memory_resource for device memory allocations
+ */
+template <aggregation::Kind op,
+          typename InputType,
+          typename PrecedingIter,
+          typename FollowingIter,
+          CUDF_ENABLE_IF(!cudf::is_fixed_width<InputType>())>
+std::unique_ptr<column> compute_lead_lag_for_nested(column_view const& input,
+                                                    column_view const& default_outputs,
+                                                    PrecedingIter preceding,
+                                                    FollowingIter following,
+                                                    size_type offset,
+                                                    rmm::cuda_stream_view stream,
+                                                    rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(default_outputs.type().id() == input.type().id(),
+               "Defaults column type must match input column.");  // Because LEAD/LAG.
+
+  CUDF_EXPECTS(default_outputs.is_empty() || (input.size() == default_outputs.size()),
+               "Number of defaults must match input column.");
+
+  // For LEAD(0)/LAG(0), no computation need be performed.
+  // Return copy of input.
+  if (offset == 0) { return std::make_unique<column>(input, stream, mr); }
+
+  // Algorithm:
+  //
+  // 1. Construct gather_map with the LEAD/LAG offset applied to the indices.
+  //    E.g. A gather_map of:
+  //        {0, 1, 2, 3, ..., N-3, N-2, N-1}
+  //    would select the input column, unchanged.
+  //
+  //    For LEAD(2), the following gather_map is used:
+  //        {3, 4, 5, 6, ..., N-1, NULL_INDEX, NULL_INDEX}
+  //    where `NULL_INDEX` selects `NULL` for the gather.
+  //
+  //    Similarly, LAG(2) is implemented using the following gather_map:
+  //        {NULL_INDEX, NULL_INDEX, 0, 1, 2...}
+  //
+  // 2. Gather input column based on the gather_map.
+  // 3. If default outputs are available, scatter contents of default_outputs`
+  //    to all positions where nulls where gathered in step 2.
+  //
+  // Note: Step 3 can be switched to use `copy_if_else()`, once it supports
+  //       nested types.
```

Comment:
can the entire algorithm use `copy_if_else(column_view(offset))` instead?

Reply:
Sadly, no. I'm afraid not. At the moment, `copy_if_else()` doesn't handle nested types.

Your idea is sound, though. I'll be addressing the addition of nested-types to `copy_if_else()` in #8135.